### PR TITLE
Tickets/dm 37343: Implement LSST Science Pipelines detection in GenerateDonutDirectDetectTask

### PR DESCRIPTION
**Pull request by jbkalmbach**

### CODE REVIEW

File: python/lsst/ts/wep/task/DonutQuickMeasurementTask.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        sources = self.detectObjectsInExp(
+            exp,
+            nSigma=self.config.nSigmaDetection,
+            nPixMin=self.config.nPixMinDetection,
+        )
+
+        fpSet = sources.getFootprints()
+        self.log.info("Found %d sources in exposure", len(fpSet))
+
+        objData = {}
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/DonutQuickMeasurementTask.py]: Maybe `objData = dict()` ?

File: python/lsst/ts/wep/task/DonutQuickMeasurementTask.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+                    and flux measurements.
+        """
+        if donutDiameter is None:
+            donutDiameter = self.config.donutDiameter
+        if cutoutPadding is None:
+            cutoutPadding = self.config.initialCutoutPadding
+
+        self.plateScale = exp.getWcs().getPixelScale().asArcseconds()
+        median = np.nanmedian(exp.image.array)
+        exp.image -= median  # is put back later
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/DonutQuickMeasurementTask.py]: Please help to add the comment to explain why you need to do `exp.image -= median` first, followed by `exp.image += median` at a latter time.

File: python/lsst/ts/wep/task/DonutQuickMeasurementTask.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        fpSet = sources.getFootprints()
+        self.log.info("Found %d sources in exposure", len(fpSet))
+
+        objData = {}
+        nMeasured = 0
+
+        exp.image.array = copy(expImageCopy)
+        self.installPsf.run(exp)
+
+        for srcNum, fp in enumerate(fpSet):
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/DonutQuickMeasurementTask.py:142]: You can put `nMeasured = 0` (line 110) right before `for srcNum, fp in enumerate(fpSet):` (line 115) to improve the coherence (or readibility) of code.

File: python/lsst/ts/wep/task/DonutQuickMeasurementTask.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+            objData[srcNum] = self._measurementResultToDict(result)
+            nMeasured += 1
+
+        self.log.info("Measured %d of %d sources in exposure", nMeasured, len(fpSet))
+
+        exp.image += median  # put background back in
+
+        boxSize = (
+            donutDiameter + cutoutPadding
+        )  # allow some slack, as cutting off side of donut is very bad
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/DonutQuickMeasurementTask.py]: Please explain a little bit why `cutting off side of donut is very bad`.

File: python/lsst/ts/wep/task/GenerateDonutDirectDetectTask.py [hunk 5]
```diff
@@ -85,42 +83,25 @@ class GenerateDonutDirectDetectTaskConfig(
...
-    """
-
-    # Config setting for pipeline task with defaults
-    donutTemplateSize = pexConfig.Field(
-        doc="Size of Template in pixels", dtype=int, default=160
+    # """
+    # Configuration settings for GenerateDonutDirectDetectTask.
+    # Specifies filter and camera details as well as subtasks
+    # that run to do the source selection.
+    # """
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/GenerateDonutDirectDetectTask.py]: I thought `"""` can do the block comment already. Why `#"""`?

File: python/lsst/ts/wep/task/GenerateDonutDirectDetectTask.py [hunk 6]
```diff
@@ -263,36 +248,45 @@ def run(
         templateMaker = DonutTemplateFactory.createDonutTemplate(
             DonutTemplateType.Model
         )
-
+        templateSize = self.config.donutDiameter + self.config.initialCutoutPadding
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/GenerateDonutDirectDetectTask.py]: If I am not wrong and you expect the `templateSize` is interger, maybe do the following is more robust:

```
templateSize = int(self.config.donutDiameter + self.config.initialCutoutPadding)
```

File: tests/task/test_generateDonutDirectDetectTask.py [hunk 7]
```diff
@@ -143,36 +142,35 @@ def testPipeline(self):
...
-        self.assertEqual(len(donutCatDf_S11), 2)
-        self.assertEqual(len(donutCatDf_S10), 2)
+        # and 1 source much brighter than blend so is "isolated"
+        self.assertEqual(len(donutCatDf_S11), 3)
+        self.assertEqual(len(donutCatDf_S10), 3)
 
         # Check outputs are correct
         outputDf = pd.concat([donutCatDf_S11, donutCatDf_S10])
-        self.assertEqual(len(outputDf), 4)
+        self.assertEqual(len(outputDf), 6)
```
Comment (teweitsai) [inline tests/task/test_generateDonutDirectDetectTask.py:151]: Just want to confirm the test result changes from 2 to 3 and 4 to 6 is expected or not.

File: tests/task/test_generateDonutDirectDetectTask.py [hunk 8]
```diff
@@ -143,36 +142,35 @@ def testPipeline(self):
...
+        tolerance = 15  # pixels
 
         result_y = np.sort(outputDf["centroid_y"].values)
-        truth_y = np.sort(np.array([3196, 2198, 2196, 3197]))
+        truth_y = np.sort(np.array([3196, 2198, 398, 2196, 3197, 398]))
         diff_y = np.sum(result_y - truth_y)
 
         result_x = np.sort(outputDf["centroid_x"].values)
-        truth_x = np.sort(np.array([3815, 2814, 2811, 3812]))
+        truth_x = np.sort(np.array([3815, 2814, 617, 2811, 3812, 614]))
```
Comment (teweitsai) [inline tests/task/test_generateDonutDirectDetectTask.py:173]: Just want to confirm, the changed values of `truth_x` and `truth_y` are expected or not.

**Review by teweitsai — APPROVED** [review]
Looks good. If the failed Jenkins test is related to the fixed UID, it might be a good chance to fix it at this PR. Thanks!

File: python/lsst/ts/wep/task/GenerateDonutDirectDetectTask.py (hunk 6 above)
Comment (jbkalmbach) [reply python/lsst/ts/wep/task/GenerateDonutDirectDetectTask.py]: Yeah, it should be an integer. I will add that.

File: tests/task/test_generateDonutDirectDetectTask.py (hunk 7 above)
Comment (jbkalmbach) [reply tests/task/test_generateDonutDirectDetectTask.py:151]: Yes, I checked this by loading the test data. Here is the image:
image

The old version was not picking up the bright, blended object on the top but now we are.

Comment (jbkalmbach) [reply tests/task/test_generateDonutDirectDetectTask.py:151]: This object is bright enough to count as "isolated" so we keep it now.

File: tests/task/test_generateDonutDirectDetectTask.py (hunk 8 above)
Comment (jbkalmbach) [reply tests/task/test_generateDonutDirectDetectTask.py:173]: Yes, see above.

**Comment by jbkalmbach** [comment]
Hi @teweitsai I added a new configuration since your review if you want to take a quick look at it. After that I think I will be ready to merge.

File: python/lsst/ts/wep/task/GenerateDonutDirectDetectTask.py [hunk 9]
```diff
@@ -263,36 +249,45 @@ def run(
...
+            donutDiameter=self.config.donutDiameter,
+            cutoutPadding=self.config.initialCutoutPadding,
         )
-
-        # make a donut catalog :
-        # 1) remove the blends in donut catalog
-        if self.config.removeBlends:
-            donutDfCopy = donutDf[~donutDf["blended"]].copy()
+        donutDf = pd.DataFrame.from_dict(objData.detectedCatalog, orient="index")
+        donutDf[f"{filterName}_flux"] = donutDf["apFlux70"]
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/GenerateDonutDirectDetectTask.py:273]: Please help to add the comment to explain this term: `apFlux70`.

**Review by teweitsai — APPROVED** [review]
Looks good! Thanks for this improvement!

**Comment by suberlak** [comment]
Is it possible for the task to read the instrument-specific configuration from the instrument policy yaml files? I wonder because needing to set all the instrument parameters whenever an instrument is changed is quite cumbersome, eg

Instead of 
```
config = GenerateDonutDirectDetectTaskConfig()
# Set auxTel parameters
config.instObscuration = 0.3525
config.instFocalLength = 21.6
config.instApertureDiameter = 1.2
config.instDefocalOffset = 32.8
config.instPixelSize = 10.0e-6
config.opticalModel = 'onAxis'
config.donutSelector.useCustomMagLimit = True
config.donutSelector.magMin = 11.
config.donutSelector.unblendedSeparation = 200
```
I'd find it easier to do 
```
config = GenerateDonutDirectDetectTaskConfig()
config.instrument = 'auxTel'
```
or 

```
config = GenerateDonutDirectDetectTaskConfig()
config.instrument = 'lsstCam'
```

**Comment by jbkalmbach** [comment]
No, not the way the existing instrument configuration setup is designed, which was to make the configurations all accessible so that if you change something in an instrument you can reconfigure the task interactively. 

We could think about things to make this easier if you want but it's a bigger design issue than what belongs here.

**Pull request merged**